microprofile/WFLY-19591_MicroProfile_OpenAPI_4.0.adoc

```diff
@@ -0,0 +1,138 @@
+ developer: pferraro
+ sme: fburgizo
+  -
+ outside-perspective:
+  -
+// If this issue tracks the promotion to a higher stability level of a previously
+// completed feature, provide the URL of the https://github.com/wildfly/wildfly-proposals/issues
+// issue that was used to track the previous feature.
+promotes:
+// This should be blank during initial development of a feature. It may be used
+// after the feature is completed if a subsequent issue is field to track promotion
+// of this feature to a higher stability level
+promoted-by:
+---
+= MicroProfile OpenAPI 4.0
+:author:            Paul Ferraro
+:email:             paul.ferraro@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+WildFly currently provides MicroProfile OpenAPI 3.1 support to applications.
+This feature request upgrades support for this specification to version 4.0 as part of version 7.0 of the MicroProfile platform specification.
+
+Most notably, this release of the MicroProfile OpenAPI specification generates documentation using version 3.1 of the OpenAPI specification.
+
+https://download.eclipse.org/microprofile/microprofile-open-api-4.0/microprofile-openapi-spec-4.0.html#release_notes_40
+
+=== User Stories
+
+This is feature is covered by existing user stories.
+
+== Issue Metadata
+
+* https://issues.redhat.com/browse/WFLY-19591[Parent issue]
+
+=== Related Issues
+
+* https://issues.redhat.com/browse/WFLY-19588
+
+=== Affected Projects or Components
+
+
+=== Other Interested Projects
+
+=== Relevant Installation Types
+
+* Traditional standalone server (unzipped or provisioned by Galleon)
+* Managed domain
+* OpenShift Source-to-Image (S2I)
+* Bootable jar
+
+== Requirements
+
+* Upgrade version of MicroProfile OpenAPI specification API to 4.0.
+* Upgrade version of SmallRye OpenAPI to 4.0.x
+
+=== Changed requirements
+
+N/A
+
+=== Non-Requirements
+
+N/A
+
+=== Future Work
+
+N/A
+
+== Backwards Compatibility
+
+=== Default Configuration
+
+N/A
+
+=== Importing Existing Configuration
+
+N/A
+
+=== Deployments
+
+See https://download.eclipse.org/microprofile/microprofile-open-api-4.0/microprofile-openapi-spec-4.0.html#incompatible_changes_40[here] for the list of incompatible changes between versions 3.1 and 4.0 of the MicroProfile OpenAPI specification.
+
+=== Interoperability
+
+N/A
+
+== Implementation Plan
+
+The implementation of MicroProfile OpenAPI 4.0 is provided by SmallRye OpenAPI 4.0.x.
+The implemetation plan consist of simple version upgrade of the following components:
+
+* org.eclipse.microprofile.openapi API
+* org.eclipse.microprofile.openapi TCK
+* io.smallrye.openapi
+
+== Admin Clients
+
+N/A
+
+== Security Considerations
+
+N/A
+
+[[test_plan]]
+== Test Plan
+
+Version 4.0 of the MicroProfile OpenAPI TCK tests will be validated as part of the WildFly testsuite.
+WildFly-specific behavior and configuration are covered by existing tests within the WildFly testsuite.
+
```

Comment:
IMO [SmallRye OpenAPI testsuite/TCKs](https://github.com/smallrye/smallrye-open-api/tree/main/testsuite) execution should be mentioned as well, since it has been so far as part of the component testing. WDYT?

Finally, since the feature is targeting the `default` stability level, it _seems_ that we should be more precise about the test categories, at least according to the [proposal template](https://github.com/wildfly/wildfly-proposals/blob/ca1db03fdb7f5073ab9d89ade44dc43b471fd919/design-doc-template.adoc#test-plan). WDYT?

Reply:
> IMO SmallRye OpenAPI testsuite/TCKs execution should be mentioned as well, since it has been so far as part of the component testing. WDYT?

Line 129.

> Finally, since the feature is targeting the default stability level, it seems that we should be more precise about the test categories, at least according to the proposal template.

I'm not sure what you mean.  There ony 2 relevant test categories here, spec-compatiblity (i.e. TCK) tests (line 129) and integration tests (line 130).

Reply:
You're right, I'll add this below as well, but the MicroProfile test suite should be taken into account as well, see https://github.com/jboss-eap-qe/eap-microprofile-test-suite.
Anyway, the main point here is that we should cover all the relevant categories that the template asks for.

Reply:
OK - I didn't know this test suite existed.  Thanks - will update to include this.